File: front/src/model/assembly/Phase.ts
import { Fingerprint } from "../Crypto";

export namespace Phase {
  export type Proposed = {
    readonly tag: "proposed";
    remaining: Fingerprint[];
  };

  export function proposed(remaining: Fingerprint[]): Proposed {
    return {
      tag: "proposed",
      remaining: remaining,
    };
  }

  export type Started = {
    readonly tag: "started";
  };

  export const started: Started = {
    tag: "started",
  };
}

export type Phase = Phase.Proposed | Phase.Started;

File: back/src/main/scala/models/assembly/Phase.scala
package chrilves.kuzh.back.models.assembly

import scala.collection.*
import chrilves.kuzh.back.models.*
import io.circe.*
import io.circe.syntax.*

enum Phase:
  case Proposed(remaining: immutable.Set[Member.Fingerprint])
  case Started(protocolState: HarvestProtocol)

object Phase:
  given phaseEncoder[A]: Encoder[Phase] with
    final def apply(s: Phase): Json =
      s match
        case Proposed(remaining) =>
          Json.obj(
            "tag"       -> "proposed".asJson,
            "remaining" -> remaining.toList.asJson
          )
        case Started(_) =>
          Json.obj(
            "tag" -> "started".asJson
          )
